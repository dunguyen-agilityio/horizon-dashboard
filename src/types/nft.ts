// Models
import { NFT } from '@/models/NFT';

export interface INFTResponse {
  id: number;
  attributes: Omit<NFT, 'id'>;
}
