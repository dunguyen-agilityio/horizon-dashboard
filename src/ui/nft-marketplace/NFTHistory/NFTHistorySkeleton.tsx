// components
import { Skeleton } from '@nextui-org/skeleton';

const NFTHistoryCardSkeleton = () => (
  <div className="flex items-center gap-4 h-[106px] px-4 py-2">
    <Skeleton className="h-[66px] w-[66px] rounded-md dark:bg-indigo-light" />
    <div className="flex flex-col gap-2 flex-1 px-6">
      <Skeleton className="h-10 w-full rounded-lg dark:bg-indigo-light" />
      <Skeleton className="h-6 w-full rounded-lg dark:bg-indigo-light" />
    </div>
  </div>
);

const NFTHistorySkeleton = () => (
  <div className="flex flex-col bg-white dark:bg-indigo rounded-md w-full 4xl:max-w-nftCard overflow-y-hidden shadow-md pb-[46px] pt-[19px] h-[605px] gap-3">
    <div className="flex items-center justify-between p-[18px]">
      <Skeleton className="dark:bg-indigo-light w-28 h-10 rounded-lg" />
      <Skeleton className="dark:bg-indigo-light w-20 h-10 rounded-lg" />
    </div>

    <div className="flex-1 flex justify-center flex-col gap-2 px-3">
      {Array(4)
        .fill(0)
        .map((_, index) => (
          <NFTHistoryCardSkeleton key={index} />
        ))}
    </div>
  </div>
);

export default NFTHistorySkeleton;
