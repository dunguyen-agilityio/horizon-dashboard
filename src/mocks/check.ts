import { TCheck } from '@/types/check';

export const MOCK_CHECKS: TCheck[] = [
  {
    id: '1',
    name: 'Tony Reichert',
    progress: 45,
    quantity: 20,
    createdAt: new Date('2024-08-14T08:25:04.188Z'),
  },
  {
    id: '2',
    name: 'Zoey Lang',
    progress: 15,
    quantity: 20,
    createdAt: new Date('2024-08-14T08:25:04.188Z'),
  },
  {
    id: '3',
    name: 'Jane Fisher',
    progress: 55,
    quantity: 28,
    createdAt: new Date('2024-08-14T08:25:04.188Z'),
  },
  {
    id: '4',
    name: 'William Howard',
    progress: 65,
    quantity: 30,
    createdAt: new Date('2024-08-14T08:25:04.188Z'),
  },
];
