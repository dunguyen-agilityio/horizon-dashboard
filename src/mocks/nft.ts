// Types
import { NFTResponse, type NFTData } from '@/models/NFT';

// Constants
import { NFT_IMAGES } from '@/constants/images';

// Mocks
import { MOCK_IMAGE_RESPONSE, MOCK_USERS, MOCK_USERS_RESPONSE } from './user';
import { StrapiResponse } from '@/types/strapi';

const [MOCK_USER] = MOCK_USERS;

export const MOCK_NFTS: NFTData[] = [
  {
    id: '1',
    name: 'Mesh Gradients',
    author: MOCK_USER,
    price: 1.23,
    image: { url: NFT_IMAGES.DEFAULT, hash: 'mock hash' },
    members: MOCK_USERS,
    createdAt: new Date('2024-08-17T07:25:04.188Z'),
  },
];

export const MOCK_NFTS_RESPONSE: StrapiResponse<NFTResponse>[] = [
  {
    id: '1',
    attributes: {
      name: 'Mesh Gradients',
      author: { data: MOCK_USERS_RESPONSE[0] },
      price: 1.23,
      members: { data: MOCK_USERS_RESPONSE },
      image: { data: MOCK_IMAGE_RESPONSE },
      createdAt: new Date('2024-08-17T07:25:04.188Z'),
    },
  },
];
