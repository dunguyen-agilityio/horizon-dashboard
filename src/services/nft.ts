import { NFTResponse } from '@/models/NFT';
import { User } from '@/models/User';
import { apiClient } from './api';
import { API_ENTITY } from '@/constants';
import { StrapiModelWithPagination, StrapiResponse } from '@/types/strapi';

type UserFavoriteResponse = User & {
  favorite: {
    id: string;
    nfts: NFTResponse[];
  } | null;
};

export const getNFTsFavorite = async () => {
  const searchParams = new URLSearchParams({
    '[populate][favorite][populate][nfts][populate][image]': 'url',
    '[populate][favorite][populate][nfts][populate][members][populate][avatar]':
      'url',
    '[populate][favorite][populate][nfts][populate][author][populate][avatar]':
      'url',
  });

  const { data, error } = await apiClient.get<UserFavoriteResponse>(
    `${API_ENTITY.USERS}/me?${searchParams.toString()}`,
    { cache: 'no-store' },
  );

  return { data: data?.favorite?.nfts || [], error };
};

type NFTsResponse = StrapiModelWithPagination<StrapiResponse<NFTResponse>[]>;

export const getNFTsRecentlyAdded = async () => {
  const searchParams = new URLSearchParams({
    'sort[0]': 'createdAt:desc',
    '[populate][image]': '*',
    '[populate][members][populate][avatar]': '*',
    '[populate][author][populate][avatar]': '*',
    '[pagination][page]': '1',
    '[pagination][pageSize]': '6',
  });

  const { data, error } = await apiClient.get<NFTsResponse>(
    `${API_ENTITY.NFTS}?${searchParams.toString()}`,
    { cache: 'no-store' },
  );

  return { data: data?.data || [], error };
};

export const getNFTsHistory = async () => {
  const searchParams = new URLSearchParams({
    'sort[0]': 'createdAt:asc',
    '[populate][image]': '*',
    '[populate][members][populate][avatar]': '*',
    '[populate][author][populate][avatar]': '*',
    '[pagination][page]': '1',
    '[pagination][pageSize]': '6',
  });

  const { data, error } = await apiClient.get<NFTsResponse>(
    `${API_ENTITY.NFTS}?${searchParams.toString()}`,
    { cache: 'no-store' },
  );

  return { data: data?.data || [], error };
};
