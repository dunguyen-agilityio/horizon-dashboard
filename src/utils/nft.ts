import { NFTDataExtend, NFTResponse } from '@/models/NFT';
import { StrapiModel, StrapiResponse } from '@/types/strapi';
import { TUser, User } from '@/models/User';
import { TImage } from '@/models/Image';

export const formatStrapiModel = <T extends { id: string }>({
  attributes,
  id,
}: StrapiResponse<T>) =>
  ({
    ...attributes,
    id,
  }) as T;

export type UserResponse1 = Omit<TUser, 'avatar' | 'role'> & {
  avatar: StrapiModel<StrapiResponse<TImage>> | null;
  role: StrapiResponse<{ name: string; id: string }> | null;
};

const formatUserResponse = (data: StrapiResponse<UserResponse1>): User => {
  const { avatar, role, ...rest } = formatStrapiModel(data);

  const { attributes: avatarData, id } = avatar?.data ?? {};

  const hasAvatar = avatarData && id;

  return new User({
    ...rest,
    role: { name: role?.attributes.name ?? '' },
    ...(hasAvatar && {
      avatar: { hash: avatarData.hash, url: avatarData.url },
    }),
  });
};

export const formatNFTResponse = ({
  id,
  attributes: { author, image, members, ...rest },
}: StrapiResponse<NFTResponse>): NFTDataExtend => {
  const { hash = '', url = '' } = image?.data?.attributes || {};

  return {
    id,
    ...rest,
    author: formatUserResponse(author.data),
    members: members.data?.map?.((member) => formatUserResponse(member)) || [],
    image: { hash, url },
  };
};
