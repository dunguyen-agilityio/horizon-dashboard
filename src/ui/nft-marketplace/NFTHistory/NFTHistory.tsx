// components
import { Text, Button } from '@/components';
import ImageWithBlur from '@/components/ImageWithBlur';

// Utils
import { formatDateRelativeToNow } from '@/utils/format';
import { NFTResponse } from '@/models/NFT';
import { StrapiResponse } from '@/types/strapi';
import { formatNFTResponse } from '@/utils/nft';

interface NFTHistoryProps {
  historyList: StrapiResponse<NFTResponse>[];
}

const NFTHistory = ({ historyList }: NFTHistoryProps) => (
  <div className="flex flex-col bg-white dark:bg-indigo rounded-md w-full 4xl:max-w-nftCard overflow-auto shadow-md pb-[46px] pt-[19px]">
    <div className="flex items-center justify-between p-3 xs:p-[18px]">
      <Text className="leading-8 font-poppins font-bold text-blue-450">
        History
      </Text>

      <Button className="text-blue-450 dark:text-white font-medium bg-gray dark:bg-purple-750 rounded-[70px]">
        See all
      </Button>
    </div>

    <div className="flex justify-center flex-col px-3 xs:px-[18px]">
      {historyList.map((nft, index) => {
        const { image, author, name, price, createdAt } =
          formatNFTResponse(nft);

        return (
          <div
            key={index}
            className="flex gap-8 justify-between p-4 xs:p-5 rounded-md hover:shadow-md hover:dark:bg-indigo-light"
          >
            <div className="flex items-center gap-4">
              <ImageWithBlur
                src={image.url}
                alt={name}
                width={66}
                height={66}
                additionalClasses="rounded-xl w-10 h-10 xs:w-[66px] xs:h-[66px] object-cover"
              />
              <div className="flex flex-col">
                <Text className="leading-tiny font-bold truncate xl:max-w-32">
                  {name}
                </Text>
                <Text className="leading-small text-small text-secondary">
                  By {author.fullName}
                </Text>
              </div>
            </div>

            <div className="flex items-center">
              <Text className="font-bold">{price} $</Text>
            </div>

            <div className="hidden xs:flex items-center">
              <Text className="text-secondary">
                {formatDateRelativeToNow(createdAt)}
              </Text>
            </div>
          </div>
        );
      })}
    </div>
  </div>
);

export default NFTHistory;
