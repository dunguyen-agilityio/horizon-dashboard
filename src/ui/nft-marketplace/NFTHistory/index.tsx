import Image from 'next/image';

// components
import { Text, Button } from '@/components';

// Utils
import { formatDateRelativeToNow } from '@/utils/format';

interface History {
  image: string;
  title: string;
  author: string;
  price: string;
  timeAgo: Date;
}

interface NFTHistoryProps {
  historyList: History[];
}

const NFTHistory = ({ historyList }: NFTHistoryProps) => (
  <div className="flex flex-col bg-white dark:bg-indigo rounded-md w-full 4xl:max-w-nftCard overflow-auto shadow-md pb-[46px] pt-[19px]">
    <div className="flex items-center justify-between p-3 xs:p-[18px]">
      <Text className="leading-8 font-poppins font-bold text-blue-450">
        History
      </Text>

      <Button className="text-blue-450 dark:text-white font-medium bg-gray dark:bg-purple-750 rounded-[70px]">
        See all
      </Button>
    </div>

    <div className="flex justify-center flex-col px-3 xs:px-[18px]">
      {historyList.map(({ image, title, author, price, timeAgo }, index) => (
        <div
          key={index}
          className="flex gap-8 justify-between p-4 xs:p-5 rounded-md hover:shadow-md hover:dark:bg-indigo-light"
        >
          <div className="flex items-center gap-4">
            <Image
              src={image}
              alt={title}
              width={66}
              height={66}
              className="rounded-xl w-10 h-10 xs:w-[66px] xs:h-[66px] object-cover"
            />
            <div className="flex flex-col">
              <Text className="leading-tiny font-bold">Colorful Heaven</Text>
              <Text className="leading-small text-small text-secondary">
                By {author}
              </Text>
            </div>
          </div>

          <div className="flex items-center">
            <Text className="font-bold">{price} $</Text>
          </div>

          <div className="hidden xs:flex items-center">
            <Text className="text-secondary">
              {formatDateRelativeToNow(timeAgo)}
            </Text>
          </div>
        </div>
      ))}
    </div>
  </div>
);

export default NFTHistory;
