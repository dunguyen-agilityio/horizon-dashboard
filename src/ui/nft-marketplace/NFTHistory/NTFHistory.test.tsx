import { render } from '@testing-library/react';

import NFTHistory from './NFTHistory';
import { MOCK_NFTS_RESPONSE } from '@/mocks/nft';

describe('NTFHistory', () => {
  it('should match snapshot', () => {
    const { container } = render(
      <NFTHistory historyList={MOCK_NFTS_RESPONSE} />,
    );
    expect(container).toMatchSnapshot();
  });

  it('should render correct title', () => {
    const { getByText } = render(
      <NFTHistory historyList={MOCK_NFTS_RESPONSE} />,
    );
    expect(getByText('History')).toBeInTheDocument();
  });
});
