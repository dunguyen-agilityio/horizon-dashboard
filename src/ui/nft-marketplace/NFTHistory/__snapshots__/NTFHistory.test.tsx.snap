// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NTFHistory should match snapshot 1`] = `
<div>
  <div
    class="flex flex-col bg-white dark:bg-indigo rounded-md max-w-nftCard w-full overflow-auto shadow-md"
  >
    <div
      class="flex items-center justify-between p-[18px]"
    >
      <p
        class="text-primary dark:text-white text-medium leading-8 font-poppins font-bold text-blue-450"
      >
        History
      </p>
      <button
        class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap subpixel-antialiased overflow-hidden tap-highlight-transparent data-[pressed=true]:scale-[0.97] outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-4 min-w-20 h-10 text-small gap-2 [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none data-[hover=true]:opacity-hover text-blue-450 dark:text-white font-medium bg-gray dark:bg-purple-750 rounded-[70px]"
        type="button"
      >
        See all
      </button>
    </div>
    <div
      class="flex justify-center flex-col"
    >
      <div
        class="max-w-[448px] flex gap-8 justify-between p-5 rounded-md hover:shadow-md hover:dark:bg-indigo-light"
      >
        <div
          class="flex items-center gap-4"
        >
          <div
            class="w-[66px] h-[66px] items-center hidden md:flex"
          >
            <img
              alt="Colorful Heaven"
              class="rounded-xl"
              data-nimg="1"
              decoding="async"
              height="66"
              loading="lazy"
              src="/_next/image?url=%2Fnft-default.webp&w=256&q=75"
              srcset="/_next/image?url=%2Fnft-default.webp&w=96&q=75 1x, /_next/image?url=%2Fnft-default.webp&w=256&q=75 2x"
              style="color: transparent;"
              width="66"
            />
          </div>
          <div
            class="flex flex-col"
          >
            <p
              class="text-primary dark:text-white text-medium leading-tiny font-bold"
            >
              Colorful Heaven
            </p>
            <p
              class="text-primary dark:text-white leading-small text-small text-secondary"
            >
              By 
              Mark Benjamin
            </p>
          </div>
        </div>
        <div
          class="flex items-center"
        >
          <p
            class="text-primary dark:text-white text-medium font-bold"
          >
            1.33
             $
          </p>
        </div>
        <div
          class="flex items-center"
        >
          <p
            class="text-primary dark:text-white text-medium text-secondary"
          >
            2 hours ago
          </p>
        </div>
      </div>
      <div
        class="max-w-[448px] flex gap-8 justify-between p-5 rounded-md hover:shadow-md hover:dark:bg-indigo-light"
      >
        <div
          class="flex items-center gap-4"
        >
          <div
            class="w-[66px] h-[66px] items-center hidden md:flex"
          >
            <img
              alt="Abstract Colors"
              class="rounded-xl"
              data-nimg="1"
              decoding="async"
              height="66"
              loading="lazy"
              src="/_next/image?url=%2Fnft-default.webp&w=256&q=75"
              srcset="/_next/image?url=%2Fnft-default.webp&w=96&q=75 1x, /_next/image?url=%2Fnft-default.webp&w=256&q=75 2x"
              style="color: transparent;"
              width="66"
            />
          </div>
          <div
            class="flex flex-col"
          >
            <p
              class="text-primary dark:text-white text-medium leading-tiny font-bold"
            >
              Colorful Heaven
            </p>
            <p
              class="text-primary dark:text-white leading-small text-small text-secondary"
            >
              By 
              Esthera Jackson
            </p>
          </div>
        </div>
        <div
          class="flex items-center"
        >
          <p
            class="text-primary dark:text-white text-medium font-bold"
          >
            0.91
             $
          </p>
        </div>
        <div
          class="flex items-center"
        >
          <p
            class="text-primary dark:text-white text-medium text-secondary"
          >
            6 hours ago
          </p>
        </div>
      </div>
      <div
        class="max-w-[448px] flex gap-8 justify-between p-5 rounded-md hover:shadow-md hover:dark:bg-indigo-light"
      >
        <div
          class="flex items-center gap-4"
        >
          <div
            class="w-[66px] h-[66px] items-center hidden md:flex"
          >
            <img
              alt="ETH AI Brain"
              class="rounded-xl"
              data-nimg="1"
              decoding="async"
              height="66"
              loading="lazy"
              src="/_next/image?url=%2Fnft-default.webp&w=256&q=75"
              srcset="/_next/image?url=%2Fnft-default.webp&w=96&q=75 1x, /_next/image?url=%2Fnft-default.webp&w=256&q=75 2x"
              style="color: transparent;"
              width="66"
            />
          </div>
          <div
            class="flex flex-col"
          >
            <p
              class="text-primary dark:text-white text-medium leading-tiny font-bold"
            >
              Colorful Heaven
            </p>
            <p
              class="text-primary dark:text-white leading-small text-small text-secondary"
            >
              By 
              Nick Wilson
            </p>
          </div>
        </div>
        <div
          class="flex items-center"
        >
          <p
            class="text-primary dark:text-white text-medium font-bold"
          >
            2.82
             $
          </p>
        </div>
        <div
          class="flex items-center"
        >
          <p
            class="text-primary dark:text-white text-medium text-secondary"
          >
            5 hours ago
          </p>
        </div>
      </div>
      <div
        class="max-w-[448px] flex gap-8 justify-between p-5 rounded-md hover:shadow-md hover:dark:bg-indigo-light"
      >
        <div
          class="flex items-center gap-4"
        >
          <div
            class="w-[66px] h-[66px] items-center hidden md:flex"
          >
            <img
              alt="Swipe Circles"
              class="rounded-xl"
              data-nimg="1"
              decoding="async"
              height="66"
              loading="lazy"
              src="/_next/image?url=%2Fnft-default.webp&w=256&q=75"
              srcset="/_next/image?url=%2Fnft-default.webp&w=96&q=75 1x, /_next/image?url=%2Fnft-default.webp&w=256&q=75 2x"
              style="color: transparent;"
              width="66"
            />
          </div>
          <div
            class="flex flex-col"
          >
            <p
              class="text-primary dark:text-white text-medium leading-tiny font-bold"
            >
              Colorful Heaven
            </p>
            <p
              class="text-primary dark:text-white leading-small text-small text-secondary"
            >
              By 
              Peter Will
            </p>
          </div>
        </div>
        <div
          class="flex items-center"
        >
          <p
            class="text-primary dark:text-white text-medium font-bold"
          >
            2.32
             $
          </p>
        </div>
        <div
          class="flex items-center"
        >
          <p
            class="text-primary dark:text-white text-medium text-secondary"
          >
            1 day ago
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
